proof/refine/ARM/IpcCancel_R.thy

```diff
@@ -2857,39 +2856,59 @@ lemma cancelAllIPC_invs'[wp]:
+  apply (fastforce intro: if_live_state_refsE)
+  done
+
+lemma cancelAllSignals_invs'_helper:
+  "\<lbrace>all_invs_but_ct_not_inQ'
+    and (\<lambda>s. \<forall>x \<in> set q. tcb_at' x s)
+    and (\<lambda>s. (\<forall>x \<in> set q. ex_nonz_cap_to' x s))\<rbrace>
+    mapM_x (\<lambda>t. do y <- setThreadState Structures_H.thread_state.Restart t;
+                        possibleSwitchTo t
+                od) q
+   \<lbrace>\<lambda>rv. all_invs_but_ct_not_inQ'\<rbrace>"
+  apply (rule mapM_x_inv_wp2)
+   apply clarsimp
+  apply (wpsimp wp: sts_st_tcb_at'_cases sts_valid_queues
+                    valid_irq_node_lift irqs_masked_lift
+                    hoare_vcg_const_Ball_lift hoare_vcg_all_lift hoare_vcg_imp_lift'
+              simp: cteCaps_of_def o_def)
+  apply (intro conjI)
+           apply (clarsimp simp: valid_tcb_state'_def global'_no_ex_cap
+                          dest!: set_mono_suffix
+                  | (drule (1) bspec, clarsimp simp: valid_pspace'_def))+
+  done
+
 lemma cancelAllSignals_invs'[wp]:
-  "\<lbrace>invs'\<rbrace> cancelAllSignals ntfn \<lbrace>\<lambda>rv. invs'\<rbrace>"
+  "cancelAllSignals ntfnPtr \<lbrace>invs'\<rbrace>"
   apply (simp add: cancelAllSignals_def)
-  apply (rule hoare_seq_ext [OF _ get_ntfn_sp'])
-  apply (case_tac "ntfnObj ntfna", simp_all)
-    apply (wp, simp)
-   apply (wp, simp)
-  apply (rule hoare_pre)
-   apply (wp rescheduleRequired_all_invs_but_ct_not_inQ
-             cancel_all_invs'_helper hoare_vcg_const_Ball_lift
-             valid_irq_node_lift ssa_invs' irqs_masked_lift
-          | simp only: sch_act_wf.simps)+
-  apply (clarsimp simp: invs'_def valid_state'_def valid_ntfn'_def)
-  sorry (*
-  apply (frule obj_at_valid_objs', clarsimp)
-  apply (clarsimp simp: projectKOs valid_obj'_def valid_ntfn'_def)
-  apply (drule(1) sym_refs_ko_atD')
-  apply (rule conjI, clarsimp elim!: if_live_state_refsE)
-  apply (rule conjI[rotated])
-   apply (clarsimp elim!: if_live_state_refsE)
-   apply (drule_tac x="(x, NTFNSignal)" in bspec)
-    apply (clarsimp simp: st_tcb_at_refs_of_rev')+
-   apply (drule st_tcb_at_state_refs_ofD')
-   apply clarsimp
-  apply (erule delta_sym_refs)
-   apply (clarsimp split: if_split_asm)
-  apply (clarsimp split: if_split_asm)
-   apply (fastforce simp: symreftype_inverse' ntfn_bound_refs'_def)
-  apply (drule_tac x="(x, NTFNSignal)" in bspec)
-   apply (clarsimp simp: st_tcb_at_refs_of_rev')+
-  apply (drule st_tcb_at_state_refs_ofD')
-  apply (fastforce simp: symreftype_inverse' ntfn_bound_refs'_def tcb_bound_refs'_def)
-  done *)
+  apply (rule hoare_seq_ext[OF _ stateAssert_sp])
+  apply (rule hoare_seq_ext[OF _ get_ntfn_sp'])
+  apply (case_tac "ntfnObj ntfn"; simp)
+    apply wpsimp
+   apply wpsimp
+  apply (wpsimp wp: rescheduleRequired_all_invs_but_ct_not_inQ sts_st_tcb_at'_cases
+                    cancelAllSignals_invs'_helper hoare_vcg_const_Ball_lift
+                    sts_valid_queues hoare_drop_imps hoare_vcg_all_lift)
```

Comment:
Could this whole block be automated to just the last `wpsimp`? It should do the case distinction automatically. It might need `wp` rule additions for `get_ntfn` (maybe, should be there by default) and `stateAssert`.

Reply:
Rearranging this turned out to be quite nasty. I'm not sure exactly why, but when the wpsimp has to do the case distinction, it struggles with the `all_invs_but_ct_not_inQ` term. This is an abbreviation, and it seems like it does some rewriting. So it'll use `rescheduleRequired_all_invs_but_ct_not_inQ`, do some rewriting, and then won't be able to apply `cancelAllSignals_invs'_helper` immediately. 

One way of solving this is to make `all_invs_but_ct_not_inQ` a definition. I'm not sure whether we want these things (`all_invs_but_blah`) to be abbreviations or definitions (we seem to have both). Another approach is to make `valid_dom_schedule'` a definition. This is listed as a FIXME. I have tried both! Making `all_invs_but_ct_not_inQ` touches far fewer lines. But making `valid_dom_schedule'` might be a better solution. Neither really make for that much of an improvement in this proof though

Reply:
Just for a bit of a laugh, I've pushed a branch to github (called `michaelm-cancelAllSignals_invs2`) that would be the change if we wanted to make `valid_dom_schedule'` a definition

Reply:
One way to avoid as much unfolding of the definition as there is would be to crunch more things for this. But considering we already unfold `invs'` and `valid_state'` all over the place, and considering that since it was an abbreviation before, simply unfolding the definition will get us back to the same proof state (in most cases), this is probably still a sensible approach

Reply:
> Just for a bit of a laugh, I've pushed a branch to github (called michaelm-cancelAllSignals_invs2) that would be the change if we wanted to make valid_dom_schedule' a definition

You mean [this bit here](https://github.com/seL4/l4v/blob/4594f5337cf4cb44f9e6688c7e34cb4d2f3f38ce/proof/refine/ARM/IpcCancel_R.thy#L2898-L2901)? 

You're right, it doesn't have much impact on this proof, and arguably makes it a bit worse because of the multiple `wpsimp` invocations without explicit `case`. Maybe let's leave this proof as is then. 

I'm unclear on whether `valid_dom_schedule'` as definition has wider impact that I haven't seen. If we're only ever just unfolding it, I'm not sure it's worth it, but I'm not against. If you feel it makes your life easier, go for it.

Reply:
Yes, that is the bit. I'm generally very much against multiple wpsimps in a row; a helper lemma is probably the only reason I'd go for it. If you'd prefer me to go back to the original proof (I would probably prefer the original proof too), then having `valid_dom_schedule'` as a definition is not necessary. Shall I just go back to the way things were in the PR originally?

Reply:
(I didn't get as far as the other change in CNodeInv, but I can look into that too)